lib/component.js

```diff
@@ -805,6 +843,91 @@ class Component extends WebComponent {
     return this._attrs;
   }
 
+  /**
+   * parse and validate config.params and create a param schema on the component
+   */
+  _initializeParams() {
+    // the real value for the params
+    this._params = {};
+    // maintain local validated map where all schema keys are defined
+    this._paramSchemas = {};
+    const paramSchemas = this.getConfig(`params`);
+    const defaultParams = this.getConfig(`defaultParams`);
+
+    for (let [paramName, paramSchema] of Object.entries(paramSchemas)) {
+      // convert type shorthand to object
+      if (!paramSchema.type) {
+        paramSchema = {type: paramSchema};
+      }
+
+      // Ensure param type is valid
+      const type = paramSchema.type;
+      if (!PARAM_TYPES.has(type)) {
+        const typeString = typeof type === `function` ? type.name : String(type);
+        throw new Error(
+          `Invalid type: ${typeString} for param: ${paramName} in paramSchema. ` +
+            `Only (${Array.from(PARAM_TYPES.keys())
+              .map((v) => `'${v.name}'`)
+              .join(` | `)}) is valid.`,
+        );
+      }
+
+      const paramSchemaObj = {
+        type,
+        required: Boolean(paramSchema.required),
+        default: defaultParams[paramName],
+      };
+
+      // set default value for the params
+      this._params[paramName] = paramSchemaObj.default;
+
+      this._paramSchemas[paramName] = paramSchemaObj;
+    }
+
+    return this._paramSchemas;
+  }
+
+  get params() {
+    return this._params;
+  }
+
+  setParams(params) {
+    const shouldComponentUpdate = this.shouldComponentUpdate(params, this.state);
+    const updateOptions = {
+      cascade: false,
+    };
+    // no extra params allowed if not defined in schema
+    for (const paramName of Object.keys(params)) {
+      if (!this._paramSchemas[paramName]) {
+        throw new Error(`extra param '${paramName}' on ${this.constructor.name} is not defined in schema`);
+      }
+    }
+
+    for (const [paramName, paramSchema] of Object.entries(this._paramSchemas)) {
+      // if param defined on schema and marked required, the key must be presented on the params
+      if (!params.hasOwnProperty(paramName) && paramSchema.required) {
+        throw new Error(
+          `param '${paramName}' on ${this.constructor.name} is defined as required param in schema but absent on component definition`,
+        );
+      }
+      const paramValue = params[paramName];
+      // set default value if empty value passed in
+      if (paramSchema.default && (paramValue === null || paramValue === undefined)) {
+        params[paramName] = paramSchema.default;
```

Comment:
this could be a bit debatable, so if a value passed in as `null` or `undefined`, we will always give it `defaultValue`.

copied from react's behavior 
https://stackblitz.com/edit/react-ts-t7fty8?file=App.tsx

Reply:
Interesting, I didn't expect that but I'm in favor of copying React's behavior.